apps/src/code-studio/pd/international_opt_in/InternationalOptIn.jsx

```diff
@@ -0,0 +1,265 @@
+import React, {PropTypes} from 'react';
+import FormController from '../form_components/FormController';
+import FormComponent from '../form_components/FormComponent';
+import DatePicker from '../workshop_dashboard/components/date_picker';
+import moment from 'moment';
+import {DATE_FORMAT} from '../workshop_dashboard/workshopConstants';
+import {
+  Row,
+  Col,
+  ControlLabel,
+  FormGroup
+} from 'react-bootstrap';
+import i18n from '@cdo/locale';
+
+export default class InternationalOptIn extends FormController {
+  static propTypes = {
+    accountEmail: PropTypes.string.isRequired,
+    labels: PropTypes.object.isRequired
+  };
+
+  /**
+   * @override
+   */
+  onSuccessfulSubmit(data) {
+    window.location = `/pd/international_opt_in/${data.id}/thanks`;
+  }
+
+  /**
+   * @override
+   */
+  serializeFormData() {
+    const formData = super.serializeFormData();
+    formData.form_data.email = this.props.accountEmail;
+    return formData;
+  }
+
+  /**
+   * @override
+   */
+  getPageComponents() {
+    return [
+      InternationalOptInComponent
+    ];
+  }
+
+  /**
+   * @override
+   */
+  getPageProps() {
+    return {
+      ...super.getPageProps(),
+      accountEmail: this.props.accountEmail,
+      labels: this.props.labels
+    };
+  }
+}
+
+
+class InternationalOptInComponent extends FormComponent {
+  static propTypes = {
+    accountEmail: PropTypes.string.isRequired
+  };
+
+  handleDateChange = (date) => {
+    // Don't allow null. If the date is cleared, default again to today.
+    date = date || moment();
+    super.handleChange({date: date.format(DATE_FORMAT)});
+  };
+
+  render() {
+    const labels = this.props.labels;
+    const date = (this.props.data && this.props.data.date) ?
+      moment(this.props.data.date, DATE_FORMAT) : moment();
+    // todo: add validation
+
+    let textFieldMapSubjects = {};
+    const lastSubjectsKey = this.props.options.subjects.slice(-1)[0];
+    textFieldMapSubjects[lastSubjectsKey] = "other";
```

Comment:
I didn't want to change the underlying form system, but it seems this would be more robust if we identified the fields based on an identifier rather than by a user-facing string, especially if we're loc'ing, like here.

Reply:
Can you elaborate? I'm not sure I understand. Do you mean here the key (option text, which does appear to be localized since it's pulled from the dynamic last value of `props.options.subjects`), or the value (which will become the key suffix for the additional text field)? If the latter, can you explain why localizing the key would be helpful? If the former, what is missing? If you want it to be a constant rather than the last in options, then by all means create a new object mapping field name to the "Other" text fields and pass it through from the server in script data.

Reply:
Oh, it's the key.  It just seemed a bit delicate to have a key in `textFieldMap` being a human-readable string (e.g. `"Other:"`, especially since it will vary with language.  As far as I can tell, this is implemented in [FormComponent](https://github.com/code-dot-org/code-dot-org/blob/f0d1496414f321d47c18ccf962bac3b483836c3d/apps/src/code-studio/pd/form_components/FormComponent.jsx#L271-L277) and I didn't want to change that code.  Hopefully not a big deal, unless there's something unexpected in any translations...

Reply:
The human-readable string is the option (out of `options` passed through from the model) that gets an additional text box. It's just used for matching purposes at render time, not stored. The right-side, the value in this mapping, becomes the suffix on the new key in the resulting `form_data` for the new field. Like other field names, this key is not (and should not be) localized.

Hopefully this explanation helps (from just above your link in `FormComponent`): 
https://github.com/code-dot-org/code-dot-org/blob/f0d1496414f321d47c18ccf962bac3b483836c3d/apps/src/code-studio/pd/form_components/FormComponent.jsx#L234-L236

If not, feel free to modify / rewrite to make it more clear.

In our case here with localized `options` the key in this mapping, the option to match, must also be localized. 

Backing up, another design choice originally would have been to have richer objects for `options` that specified aspects of the UI including whether that particular option had an additional text box. However we added this feature after options was already just a simple array, so we chose this approach instead - to specify additional text fields as an add-on, with a mapping from the options that already existed.

Anyway, if both `options` and the source for the `textFieldMap` both come from the model and are localized, then what is delicate? It's only really different from the above approach in that it's distributed over 2 lists.

Reply:
We discussed offline.  My main concern was having JS hash key names be user-facing strings that are localised, meaning they can contain special & international characters.  However, we believe that all of those are valid, so this should be fine.